File: Day #844/solution.go
// Day 844: 2-SAT via implication graph + Kosaraju SCC.
// Clause (x OR y) adds !x->y and !y->x. SAT iff no var shares an SCC with its negation. O(V+E).
package main

import "fmt"

type twoSat struct {
	n   int
	adj [][]int
}

func newTwoSat(n int) *twoSat { return &twoSat{n: n, adj: make([][]int, 2*n)} }

func node(v int, neg bool) int {
	if neg {
		return 2*v + 1
	}
	return 2 * v
}

func (t *twoSat) addClause(va int, na bool, vb int, nb bool) {
	a, b := node(va, na), node(vb, nb)
	t.adj[a^1] = append(t.adj[a^1], b)
	t.adj[b^1] = append(t.adj[b^1], a)
}

func (t *twoSat) solve() ([]bool, bool) {
	n2 := 2 * t.n
	vis := make([]bool, n2)
	order := []int{}
	var dfs1 func(u int)
	dfs1 = func(u int) {
		vis[u] = true
		for _, v := range t.adj[u] {
			if !vis[v] {
				dfs1(v)
			}
		}
		order = append(order, u)
	}
	for i := 0; i < n2; i++ {
		if !vis[i] {
			dfs1(i)
		}
	}
	radj := make([][]int, n2)
	for u := 0; u < n2; u++ {
		for _, v := range t.adj[u] {
			radj[v] = append(radj[v], u)
		}
	}
	comp := make([]int, n2)
	for i := range comp {
		comp[i] = -1
	}
	c := 0
	var dfs2 func(u, cc int)
	dfs2 = func(u, cc int) {
		comp[u] = cc
		for _, v := range radj[u] {
			if comp[v] == -1 {
				dfs2(v, cc)
			}
		}
	}
	for i := n2 - 1; i >= 0; i-- {
		u := order[i]
		if comp[u] == -1 {
			dfs2(u, c)
			c++
		}
	}
	assign := make([]bool, t.n)
	for v := 0; v < t.n; v++ {
		if comp[2*v] == comp[2*v+1] {
			return nil, false
		}
		assign[v] = comp[2*v] > comp[2*v+1]
	}
	return assign, true
}

func tf(b bool) string {
	if b {
		return "True"
	}
	return "False"
}

func main() {
	ts := newTwoSat(3) // a=0,b=1,c=2
	ts.addClause(2, true, 1, false)  // ¬c OR b
	ts.addClause(1, false, 2, false) // b OR c
	ts.addClause(1, true, 2, false)  // ¬b OR c
	ts.addClause(2, true, 0, true)   // ¬c OR ¬a
	a, ok := ts.solve()
	if !ok {
		fmt.Println("False")
		return
	}
	fmt.Printf("a = %s, b = %s, c = %s\n", tf(a[0]), tf(a[1]), tf(a[2]))
}

File: Day #844/solution.cpp
// Day 844: 2-SAT via implication graph + Tarjan SCC.
// Each clause (x OR y) adds edges !x->y and !y->x. Satisfiable iff no var shares an SCC with its negation.
// O(V+E). Variables 0..n-1; literal encoded as 2*v (true) / 2*v+1 (false/negated).
#include <bits/stdc++.h>
using namespace std;

struct TwoSat {
    int n;
    vector<vector<int>> adj;
    TwoSat(int vars): n(vars), adj(2*vars) {}
    int node(int var, bool neg){ return 2*var + (neg ? 1 : 0); }
    // clause: (lit a) OR (lit b); a/b given as (var, isNegated)
    void addClause(int va, bool na, int vb, bool nb){
        int a = node(va, na), b = node(vb, nb);
        adj[a^1].push_back(b); // !a -> b
        adj[b^1].push_back(a); // !b -> a
    }
    vector<int> comp, order, val; vector<bool> vis;
    void dfs1(int u){ vis[u]=true; for(int v: adj[u]) if(!vis[v]) dfs1(v); order.push_back(u); }
    vector<vector<int>> radj;
    void dfs2(int u, int c){ comp[u]=c; for(int v: radj[u]) if(comp[v]==-1) dfs2(v,c); }
    bool solve(vector<bool>& assign){
        vis.assign(2*n,false); order.clear();
        for(int i=0;i<2*n;i++) if(!vis[i]) dfs1(i);
        radj.assign(2*n,{});
        for(int u=0;u<2*n;u++) for(int v: adj[u]) radj[v].push_back(u);
        comp.assign(2*n,-1); int c=0;
        for(int i=2*n-1;i>=0;i--){ int u=order[i]; if(comp[u]==-1) dfs2(u,c++); }
        assign.assign(n,false);
        for(int v=0;v<n;v++){
            if(comp[2*v]==comp[2*v+1]) return false;
            assign[v] = comp[2*v] > comp[2*v+1]; // true literal has later SCC order
        }
        return true;
    }
};

int main(){
    // (¬c OR b) AND (b OR c) AND (¬b OR c) AND (¬c OR ¬a); vars a=0,b=1,c=2
    TwoSat ts(3);
    ts.addClause(2,true, 1,false);  // ¬c OR b
    ts.addClause(1,false,2,false);  // b OR c
    ts.addClause(1,true, 2,false);  // ¬b OR c
    ts.addClause(2,true, 0,true);   // ¬c OR ¬a
    vector<bool> assign;
    if(ts.solve(assign)){
        const char* nm = "abc";
        cout << "a = " << (assign[0]?"True":"False")
             << ", b = " << (assign[1]?"True":"False")
             << ", c = " << (assign[2]?"True":"False") << "\n";
    } else cout << "False\n";
    return 0;
}
